File: slang-checks/src/main/java/org/sonarsource/slang/checks/VariableAndParameterNameCheck.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.slang.checks;

import java.util.regex.Pattern;
import org.sonar.check.Rule;
import org.sonar.check.RuleProperty;
import org.sonarsource.slang.api.FunctionDeclarationTree;
import org.sonarsource.slang.api.IdentifierTree;
import org.sonarsource.slang.api.ParameterTree;
import org.sonarsource.slang.api.VariableDeclarationTree;
import org.sonarsource.slang.checks.api.CheckContext;
import org.sonarsource.slang.checks.api.InitContext;
import org.sonarsource.slang.checks.api.SlangCheck;
import org.sonarsource.slang.checks.utils.Language;
import org.sonarsource.slang.checks.utils.PropertyDefaultValue;

@Rule(key = "S117")
public class VariableAndParameterNameCheck implements SlangCheck {

  private static final String DEFAULT_FORMAT = "^[_a-z][a-zA-Z0-9]*$";

  @RuleProperty(
    key = "format",
    description = "Regular expression used to check the names against."
  )
  @PropertyDefaultValue(language = Language.KOTLIN, defaultValue = Language.KOTLIN_PARAMETERS_AND_VARIABLE_NAMING)
  @PropertyDefaultValue(language = Language.RUBY, defaultValue = Language.RUBY_NAMING_DEFAULT)
  @PropertyDefaultValue(language = Language.SCALA, defaultValue = Language.SCALA_NAMING_DEFAULT)
  @PropertyDefaultValue(language = Language.GO, defaultValue = Language.GO_NAMING_DEFAULT)
  public String format = DEFAULT_FORMAT;

  @Override
  public void initialize(InitContext init) {
    Pattern pattern = Pattern.compile(format);

    init.register(VariableDeclarationTree.class, (ctx, tree) -> {
      if (ctx.ancestors().stream().anyMatch(FunctionDeclarationTree.class::isInstance)) {
        check(pattern, ctx, tree.identifier(), "local variable");
      }
    });

    init.register(FunctionDeclarationTree.class, (ctx, tree) ->
      tree.formalParameters().stream()
        .filter(ParameterTree.class::isInstance)
        .map(ParameterTree.class::cast)
        .forEach(
        param -> check(pattern, ctx, param.identifier(), "parameter")));
  }

  private void check(Pattern pattern, CheckContext ctx, IdentifierTree identifier, String variableKind) {
    if (!pattern.matcher(identifier.name()).matches()) {
      String message = String.format("Rename this %s to match the regular expression \"%s\".", variableKind, this.format);
      ctx.reportIssue(identifier, message);
    }
  }

}

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/VariableAndParameterNameCheck.kt
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.checks

import org.jetbrains.kotlin.cfg.pseudocode.containingDeclarationForPseudocode
import org.jetbrains.kotlin.psi.KtFunction
import org.jetbrains.kotlin.psi.KtLambdaExpression
import org.jetbrains.kotlin.psi.KtNamedDeclaration
import org.jetbrains.kotlin.psi.KtNamedFunction
import org.jetbrains.kotlin.psi.KtPrimaryConstructor
import org.jetbrains.kotlin.psi.KtProperty
import org.jetbrains.kotlin.psi.KtSecondaryConstructor
import org.sonar.api.rule.RuleKey
import org.sonar.check.Rule
import org.sonar.check.RuleProperty
import org.sonarsource.kotlin.api.AbstractCheck
import org.sonarsource.kotlin.plugin.KotlinFileContext

/**
 * Replacement for [org.sonarsource.slang.checks.VariableAndParameterNameCheck]
 */
@Rule(key = "S117")
class VariableAndParameterNameCheck : AbstractCheck() {
    companion object {
        const val DEFAULT_FORMAT = "^`?[_a-z][a-zA-Z0-9]*`?$"
    }

    @RuleProperty(
        key = "format",
        description = "Regular expression used to check the names against.",
        defaultValue = DEFAULT_FORMAT)
    var format: String = DEFAULT_FORMAT

    private lateinit var formatRegex: Regex

    override fun initialize(ruleKey: RuleKey) {
        super.initialize(ruleKey)
        formatRegex = Regex(format)
    }

    override fun visitPrimaryConstructor(constructor: KtPrimaryConstructor, kotlinFileContext: KotlinFileContext) {
        constructor.valueParameters.forEach { check("parameter", it, kotlinFileContext) }
    }

    override fun visitSecondaryConstructor(constructor: KtSecondaryConstructor, kotlinFileContext: KotlinFileContext) {
        constructor.valueParameters.forEach { check("parameter", it, kotlinFileContext) }
    }

    override fun visitNamedFunction(function: KtNamedFunction, kotlinFileContext: KotlinFileContext) {
        if (function.receiverTypeReference != null) {
            /** see [org.sonarsource.kotlin.converter.KotlinTreeVisitor.createFunctionDeclarationTree] */
            return
        }
        function.valueParameters.forEach { check("parameter", it, kotlinFileContext) }
    }

    override fun visitProperty(property: KtProperty, kotlinFileContext: KotlinFileContext) {
        val containingDeclarationForPseudocode = property.containingDeclarationForPseudocode
        if (containingDeclarationForPseudocode is KtFunction && containingDeclarationForPseudocode.receiverTypeReference != null) {
            /** see [org.sonarsource.kotlin.converter.KotlinTreeVisitor.createFunctionDeclarationTree] */
            return
        }
        /** see `hasDelegate` in [org.sonarsource.kotlin.converter.KotlinTreeVisitor.createVariableDeclaration] */
        if (property.isLocal && !property.hasDelegate()) {
            check("local variable", property, kotlinFileContext)
        }
    }

    override fun visitLambdaExpression(expression: KtLambdaExpression, kotlinFileContext: KotlinFileContext) {
        expression.valueParameters.forEach { check("parameter", it, kotlinFileContext) }
    }

    private fun check(kind: String, declaration: KtNamedDeclaration, kotlinFileContext: KotlinFileContext) {
        val name = declaration.nameIdentifier?.text ?: return
        if (!name.matches(formatRegex)) {
            kotlinFileContext.reportIssue(
                declaration.nameIdentifier!!,
                "Rename this $kind to match the regular expression \"$format\".")
        }
    }

}
